taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -468,53 +471,83 @@ func (self *Contract) Run(ctx vm.CallFrame, evm *vm.EVM) ([]byte, error) {
+		return method.Outputs.Pack(self.getDelegations(args))
 
 	case "getUndelegations":
 		var args sol.GetUndelegationsArgs
 		if err = method.Inputs.Unpack(&args, input); err != nil {
 			fmt.Println("Unable to parse getUndelegations input args: ", err)
 			return nil, err
 		}
-
-		res := self.getUndelegations(args)
-		return method.Outputs.Pack(res.Undelegations, res.End)
+		return method.Outputs.Pack(self.getUndelegations(args))
 	}
 
 	return nil, nil
 }
 
-func (self *Contract) DistributeRewards(rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
+func (self *Contract) DistributeRewards(author *common.Address, rewardsStats *rewards_stats.RewardsStats, feesRewards *FeesRewards) {
 	// When calling DistributeRewards, internal structures must be always initialized
 	self.lazy_init()
 
 	// Calculates number of tokens to be generated as block reward
 	blockReward := bigutil.Mul(self.amount_delegated, self.yield_percentage)
 	blockReward = bigutil.Div(blockReward, bigutil.Mul(Big100, self.blocks_per_year))
 
+	votesReward := bigutil.Big0
+	authorReward := bigutil.Big0
+	trxsReward := blockReward
+	// We need to handle case for block 1
+	if rewardsStats.TotalVotesWeight > 0 {
+		// Calculate propotion between votes and transactions
+		trxsReward = bigutil.Div(bigutil.Mul(blockReward, VotesToTrasnactionsRatio), Big100)
+		votesReward = bigutil.Sub(blockReward, trxsReward)
+
+		// In case block author included more than minimum required number of votes
+		if rewardsStats.BonusVotesWeight > 0 && self.cfg.MaxBlockAuthorReward > 0 {
+			maxBonusVotes := rewardsStats.MaxVotesWeight / 3
+			authorRewardRatio := rewardsStats.BonusVotesWeight * uint64(self.cfg.MaxBlockAuthorReward) / maxBonusVotes
+			validatorVoteReward := bigutil.Mul(big.NewInt(int64(rewardsStats.BonusVotesWeight)), votesReward)
+			validatorVoteReward = bigutil.Div(validatorVoteReward, big.NewInt(int64(rewardsStats.TotalVotesWeight)))
+			authorReward = bigutil.Div(bigutil.Mul(validatorVoteReward, big.NewInt(int64(authorRewardRatio))), Big100)
+			votesReward = bigutil.Sub(votesReward, authorReward)
+		}
+	}
+
 	totalUniqueTxsCountCheck := uint32(0)
+	totalVoteWeightCheck := uint64(0)
+	totalRewardCheck := bigutil.Big0
+	// Add reward to the author for additional included votes
+	if authorReward.Cmp(bigutil.Big0) == 1 {
+		validator := self.validators.GetValidator(author)
+		if validator == nil {
+			panic("DistributeRewards - non existent block author")
+		}
+
+		validatorCommission := bigutil.Div(bigutil.Mul(authorReward, big.NewInt(int64(validator.Commission))), Big10000)
+		delegatorsRewards := bigutil.Sub(authorReward, validatorCommission)
+
+		validator.CommissionRewardsPool = bigutil.Add(validator.CommissionRewardsPool, validatorCommission)
+		validator.RewardsPool = bigutil.Add(validator.RewardsPool, delegatorsRewards)
+
+		totalRewardCheck = bigutil.Add(totalRewardCheck, authorReward)
+		self.validators.ModifyValidator(author, validator)
```

Comment:
little bit confusing because of naming... Would be nice to keep naming it consistent with the rest of the code, `author` should be `block_author_addr` and validator here would be `block_author`